strings/base_collections_base.h

```diff
@@ -70,43 +140,66 @@ WINRT_EXPORT namespace winrt
 
             T Current() const
             {
-                if (m_current == m_end)
+                return m_owner->perform_shared([&]
                 {
-                    throw hresult_out_of_bounds();
-                }
+                    this->check_version(*m_owner);
 
-                if constexpr (!impl::is_key_value_pair<T>::value)
-                {
-                    return m_owner->unwrap_value(*m_current);
-                }
-                else
-                {
-                    return make<impl::key_value_pair<T>>(m_owner->unwrap_value(m_current->first), m_owner->unwrap_value(m_current->second));
-                }
+                    if (m_current == m_end)
+                    {
+                        throw hresult_out_of_bounds();
+                    }
+
+                    return current_value();
+                });
             }
 
-            bool HasCurrent() const noexcept
+            bool HasCurrent() const
             {
-                return m_current != m_end;
+                return m_owner->perform_shared([&]
+                {
+                    this->check_version(*m_owner);
+                    return m_current != m_end;
+                });
             }
 
-            bool MoveNext() noexcept
+            bool MoveNext()
             {
-                if (m_current != m_end)
+                return m_owner->perform_exclusive([&]
                 {
-                    ++m_current;
-                }
+                    this->check_version(*m_owner);
+                    if (m_current != m_end)
+                    {
+                        ++m_current;
+                    }
 
-                return HasCurrent();
+                    return m_current != m_end;
+                });
             }
 
             uint32_t GetMany(array_view<T> values)
             {
-                return GetMany(values, typename std::iterator_traits<iterator_type>::iterator_category());
+                return m_owner->perform_exclusive([&]
+                {
+                    this->check_version(*m_owner);
+                    return GetMany(values, typename std::iterator_traits<iterator_type>::iterator_category());
+                });
             }
 
         private:
 
+            T current_value() const
```

Comment:
Should there be a naming convention for "no locks in this method"?  `current_value` seems to only be called under a lock.  (Not sure where we stand on SAL, but there's a "require no locks held" annotation.)

Reply:
We don't use SAL inside C++/WinRT. We do however adopt standard C++ attributes that are slowly taking the place of SAL.

Reply:
Ok, no SAL, but maybe "current_value_withlock()" ?